Initial development

paulkahle opened the pull request.

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+@import '../node_modules/suprematism-style-guide/dist/utils/_colors.scss';
```
You can short cut this to: `utils/_colors.scss`.

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+
+      position: absolute;
+
+      z-index:2;
```
Not sure if you want to use the style guide z-indexes or not.

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+
+
+        /deep/ .header{
+          font-size: 10px;
```
Should we use the style guide settings for this? We have text sizes:
https://cinbcuniversal.github.io/suprematism-style-guide/section/typography#type-placeholders

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+        }
+
+        &:after {
+          border-color: rgba(136, 183, 213, 0);
```
Are we not using the colors in the style guide?

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+
+        &:after, &:before{
+          left:100%;
+          top: 5px;
```
We have some of this setup in the style guide.

caffeinelogic commented on src/dropdown-menu.component.scss:
```diff
@@ -0,0 +1,340 @@
+@import '../node_modules/suprematism-style-guide/dist/utils/_colors.scss';
+
```
Might be worth seeing what values can be pulled from the style guide for this ...

caffeinelogic replied:
This comment is geared towards the entire scss file

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+   *
+   * stops event propogation of events that should not close the dropdown
+   */
+@Directive({ selector: '[supreCloseEvent]' })
```
Best practice per the angular style guide is to make this its own file. I would recommend pulling this into its own file.

Also wondering if this directive selector is too ambiguous? It will prevent bubbling for anything its on, not just this menu system. Should we pull this into the main app?

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+   */
+@Directive({ selector: '[supreCloseEvent]' })
+export class DropdownCloseDirective {
+  @Input('supreCloseEvent') supreCloseEvent: string;
```
You can omit the argument on the decorator: `@Input()`

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+
+  constructor(private _elementRef: ElementRef) {}
+
+  ngOnInit() {
```
You can remove this, you're not using it. Then you can remove the `implements` in the class and the `import` for it.

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+    return fits.bottom && fits.right && fits.left && fits.top;
+  }
+
+  constructor(private _elementRef: ElementRef) {}
```
Suggest that this goes before the first method in the class so a user knows what is being Injected into this class. 

Also, no need for an `_`, we have the private keyword now in TS.

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+  /** Inputs */
+  @Input('trigger') trigger = 'click';
+  @Input('pointed') pointed = true;
+  @Input('positions') positions = ['bottom-right', 'bottom-left', 'top-right', 'top-left'];
```
This is a perfect candidate for a `type`.
```
type Position = 'top-right' | 'top-middle' | 'top-left' | 'bottom-right' | 'bottom-middle' | 'bottom-left' | 'right-bottom' | 'right-middle' | 'right-top' | 'left-top' | 'left-middle' | 'left-bottom';
```

This would help ensure that the values you expect are used. The IDE will help stop typo's etc.

Then this line can be:
```
@Input() positions: Position[]  = ['bottom-right', 'bottom-left', 'top-right', 'top-left'];
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+export class DropdownMenuComponent implements OnInit {
+
+  /** Inputs */
+  @Input('trigger') trigger = 'click';
```
This is a perfect candidate for a Type:
```
type Trigger = 'hover' | 'click';
```

Then this line becomes:
```
@Input() trigger: Trigger = 'click';
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+
+  /** Inputs */
+  @Input('trigger') trigger = 'click';
+  @Input('pointed') pointed = true;
```
Can drop the argument for the `@Input()`.

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+  handleClick(event) {
+    if ( !this.isOpen ) {
+      // add listener to close dropdown when clicking elsewhere on screen
+      document.addEventListener('click', this.handleOffClick.bind(this));
```
It's generally bad practice to directly call the `document` in the class since Angular is designed to be platform dynamic. There are other ways to solve this, we can chat about some of the options that would make this more testable etc.

I also think this causes a lot of noise when I tested it. The event fired all the time.

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+
+  private isOpen = false;
+  private isPlaced = false;
+  private delay: any = { close : 1 };
```
To get rid of using `any` (which is usually a good thing) you can do the following:
```
private delay: {close: number} = { close : 1 };
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+      return;
+    }
+
+    this.menu.nativeElement.className = 'menu-activation ' + (this.pointed ? 'pointed ' : '') + position.toLowerCase();
```
You can use a template string here:
```
this.menu.nativeElement.className = `menu-activation ${(this.pointed ? 'pointed ' : '')} ${position.toLowerCase()}`;
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+
+    this.menu.nativeElement.className = 'menu-activation ' + (this.pointed ? 'pointed ' : '') + position.toLowerCase();
+    const rect = this.menu.nativeElement.getBoundingClientRect();
+    const viewport: any = {
```
This can be written like this:
```
viewport: {height: number, width: number} = ...
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+      height : Math.max(document.documentElement.clientHeight, window.innerHeight),
+      width : Math.max(document.documentElement.clientWidth, window.innerWidth)
+    };
+    const fits: any = {
```
Can be written like this:
```
fits: {bottom: boolean, right: boolean, left: boolean, top: boolean} = ...
```

caffeinelogic commented on src/dropdown-menu.component.ts:
```diff
@@ -0,0 +1,179 @@
+     * handles mouseleave event to close dropdown when trigger  = "hover"
+     * @memberOf DropdownMenuComponent
+     */
+  handleMouseLeave() {
```
I am assuming that the methods without the `public` keyword are private? If that's the case add the `private` keyword. If they are public they can be accessed from outside, and marking them as `private` means that the template can still use them, but they are not part of the API.

caffeinelogic commented on example/app/app.component.ts:
```diff
@@ -0,0 +1,9 @@
+
+@Component({
+  selector: 'supre-root',
+  template: require('./app.component.html')
```
You don't need `require` here.

paulkahle commented:
@caffeinelogic 
I've addressed all the comments. 
Regarding some of the more significant issues: I'm no longer stopping event propagation but rather setting a property on the event which the dropdown component then looks for when deciding if it should close.  
I also fixed the document click handler to only fire when the dropdown is open.  The problem was that I was binding to the even handler `this.handleOffClick.bind(this)` in both the addEventListener and removeEventListener calls creating different references.  Because the reference in removeEventLister didn't match the one in addEventListener the listener wasn't being removed.

caffeinelogic commented on src/dropdown-menu-close.directive.ts:
```diff
@@ -0,0 +1,23 @@
+    if (!event || !event.type) {
+      return;
+    }
+    event.preventSupreDropdownClose = this.supreCloseEvent !== event.type;
```
Tricky tricky 😄 

So I assume the way this bubbles up that by the time the parent gets it, that property exists?

paulkahle replied:
Yes exactly.

paulkahle commented:
@caffeinelogic 
I added the ngOnDestroy hook to remove the document click handler if it exists.
